examples/train_model.py

```diff
@@ -33,36 +36,109 @@ def run_eval(agent, opt, datatype):
         if valid_world.epoch_done():
             break
     valid_world.shutdown()
-    metrics = datatype + ":" + str(valid_world.report())
+    valid_report = valid_world.report()
+    metrics = datatype + ":" + str(valid_report)
     print(metrics)
-    # Write out metrics
-    if opt['model_file']:
-        f = open(opt['model_file'] + '.' + datatype, "a+")
-        f.write(metrics + '\n')
-        f.close()
+    if still_training:
+        return valid_report
+    else:
+        if opt['model_file']:
+            # Write out metrics
+            f = open(opt['model_file'] + '.' + datatype, "a+")
+            f.write(metrics + '\n')
+            f.close()
+
+def build_dict(opt):
+    print('[setting up dictionary.]')
+    if 'dict_loadpath' not in opt:
+        if '.model' in opt['model_file']:
+            dict_fn = opt['model_file'].replace('.model', '.dict')
+        else:
+            dict_fn = opt['model_file'] + '.dict'
+        opt['dict_loadpath'] = dict_fn
+    if os.path.isfile(opt['dict_loadpath']):
+        # dict already built
+        print("[dict already built.]")
+        return
+    opt['dict_savepath'] = opt['dict_loadpath']
+    opt.pop('dict_loadpath', None)
+    dictionary = DictionaryAgent(opt)
+    ordered_opt = copy.deepcopy(opt)
+    cnt = 0
+    for datatype in ['train:ordered', 'valid']:
+        # we use train and valid sets to build dictionary
+        ordered_opt['datatype'] = datatype
+        ordered_opt['numthreads'] = 1
+        ordered_opt['batchsize'] = 1
+        world_dict = create_task(ordered_opt, dictionary)
+        # pass examples to dictionary
+        for _ in world_dict:
+            cnt += 1
+            if cnt > opt['dict_maxexs'] and opt['dict_maxexs'] > 0:
+                print('Processed {} exs, moving on.'.format(
+                      opt['dict_maxexs']))
+                # don't wait too long...
+                break
+            world_dict.parley()
+    dictionary.save(dict_fn, sort=True)
+    opt['dict_loadpath'] = opt['dict_savepath']
+    opt.pop('dict_savepath', None)
 
 def main():
     # Get command line arguments
     parser = ParlaiParser(True, True)
     parser.add_argument('-d', '--display-examples',
                         type='bool', default=False)
-    parser.add_argument('-e', '--num-epochs', default=1)
+    parser.add_argument('-e', '--num-epochs', type=int, default=1)
     parser.add_argument('-mtt', '--max-train-time',
                         type=float, default=float('inf'))
+    parser.add_argument('-lt', '--log-every-n-secs',
+                        type=float, default=1)
+    parser.add_argument('-vt', '--validate-every-n-secs',
```

Comment:
-vt would be a great param for "validation task" in case you want to validate on something else :)